src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -15,14 +15,70 @@ private BoundExpression BindIsPatternExpression(IsPatternExpressionSyntax node,
         {
             var expression = BindExpression(node.Expression, diagnostics);
             var hasErrors = node.HasErrors || IsOperandErrors(node, expression, diagnostics);
-            var pattern = BindPattern(node.Pattern, expression, expression.Type, hasErrors, diagnostics);
+            var pattern = BindPattern(node.Pattern, expression, ExpressionIsNull(expression), expression.Type, hasErrors, diagnostics);
             return new BoundIsPatternExpression(
                 node, expression, pattern, GetSpecialType(SpecialType.System_Boolean, diagnostics, node), hasErrors);
         }
 
+        /// <summary>
+        /// Returns true if the expression is known to always be null.
+        /// Returns false if the expression is always known to be non-null.
+        /// Returns null otherwise.
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns></returns>
+        internal static bool? ExpressionIsNull(BoundExpression expression)
+        {
+            if (expression.ConstantValue != null)
+            {
+                return expression.ConstantValue.IsNull;
+            }
+
+            if (expression.Type == null || expression.Type.IsNonNullableValueType())
+            {
+                return false;
+            }
+
+            switch (expression.Kind)
+            {
+                case BoundKind.Conversion:
+                    {
+                        var conversion = (BoundConversion)expression;
+                        switch(conversion.ConversionKind)
+                        {
+                            case ConversionKind.AnonymousFunction:
+                            case ConversionKind.Boxing:
+                            case ConversionKind.ImplicitNullable:
+                            case ConversionKind.InterpolatedString:
+                            case ConversionKind.MethodGroup:
+                                return false;
+                            case ConversionKind.ExplicitReference:
+                            case ConversionKind.Identity:
+                            case ConversionKind.ImplicitReference:
+                                return ExpressionIsNull(conversion.Operand);
+                            case ConversionKind.NullLiteral:
+                                return true;
+                            default:
+                                return null;
+                        }
+                    }
+                case BoundKind.AnonymousObjectCreationExpression:
+                case BoundKind.ArrayCreation:
+                case BoundKind.DelegateCreationExpression:
+                case BoundKind.DynamicObjectCreationExpression:
+                case BoundKind.NoPiaObjectCreationExpression:
+                case BoundKind.NullCoalescingOperator:
```

Comment:
good catch.

---
In reply to: [64670654](https://github.com/dotnet/roslyn/pull/11485#discussion_r64670654) [](ancestors = 64670654)